src/tof/model.py

```diff
@@ -31,18 +51,71 @@ class Model:
 
     def __init__(
         self,
-        choppers: Union[Chopper, List[Chopper]],
-        detectors: Union[Detector, List[Detector]],
         pulse: Pulse,
+        choppers: Optional[
+            Union[Chopper, List[Chopper], Tuple[Chopper, ...], Dict[str, Chopper]]
+        ] = None,
+        detectors: Optional[
+            Union[Detector, List[Detector], Tuple[Detector, ...], Dict[str, Detector]]
+        ] = None,
     ):
-        self.choppers = choppers
-        if not isinstance(self.choppers, (list, tuple)):
-            self.choppers = [self.choppers]
-        self.detectors = detectors
-        if not isinstance(self.detectors, (list, tuple)):
-            self.detectors = [self.detectors]
+        self.choppers = _input_to_dict(choppers)
+        self.detectors = _input_to_dict(detectors)
         self.pulse = pulse
 
+    def add(self, component):
+        """
+        Add a component to the instrument.
+        Component names must be unique across choppers and detectors.
+
+        Parameters
+        ----------
+        component:
+            A chopper or detector.
+        """
+        if component.name in chain(self.choppers, self.detectors):
+            raise KeyError(
+                f"Component with name {component.name} already exists. "
+                "If you wish to replace/update an existing component, use "
+                "``model.choppers['name'] = new_chopper`` or "
+                "``model.detectors['name'] = new_detector``."
+            )
+        if isinstance(component, Chopper):
+            self.choppers[component.name] = component
+        elif isinstance(component, Detector):
+            self.detectors[component.name] = component
+        else:
+            raise TypeError(
+                f"Cannot add component of type {type(component)} to the model."
+            )
+
+    def remove(self, name: str):
+        """
+        Remove a component.
+
+        Parameters
+        ----------
+        name:
+            The name of the component to remove.
+        """
+        if name in self.choppers:
+            del self.choppers[name]
+        elif name in self.detectors:
+            del self.detectors[name]
+        else:
+            raise KeyError(f"No component with name {name} was found.")
+
+    def __iter__(self):
+        return chain(self.choppers, self.detectors)
+
+    def __getitem__(self, name) -> Union[Chopper, Detector]:
+        if name not in self:
+            raise KeyError(f"No component with name {name} was found.")
+        return self.choppers.get(name, self.detectors.get(name))
```

Comment:
Note that this gets from `detectors` even if the `choppers.get` succeeds.

Reply:
:+1:  I changed to
```Py
return self.chopper[name] if name in self.choppers else self.detectors[name]
```